lib/ansible/utils/vars.py

```diff
@@ -92,32 +92,92 @@ def combine_vars(a, b):
         return result
 
 
-def merge_hash(a, b):
+def merge_hash(x, y, recursive=True, list_merge='replace'):
     """
-    Recursively merges hash b into a so that keys from b take precedence over keys from a
+    Return a new dictionary result of the merges of y into x,
+    so that keys from y take precedence over keys from x.
+    (x and y aren't modified)
     """
-
-    _validate_mutable_mappings(a, b)
-
-    # if a is empty or equal to b, return b
-    if a == {} or a == b:
-        return b.copy()
-
-    # if b is empty the below unfolds quickly
-    result = a.copy()
-
-    # next, iterate over b keys and values
-    for k, v in iteritems(b):
-        # if there's already such key in a
-        # and that key contains a MutableMapping
-        if k in result and isinstance(result[k], MutableMapping) and isinstance(v, MutableMapping):
-            # merge those dicts recursively
-            result[k] = merge_hash(result[k], v)
-        else:
-            # otherwise, just copy the value from b to a
-            result[k] = v
-
-    return result
+    if list_merge not in ('replace', 'keep', 'append', 'prepend', 'append_rp', 'prepend_rp'):
+        raise AnsibleError("merge_hash: 'list_merge' argument can only be equal to 'replace', 'keep', 'append', 'prepend', 'append_rp' or 'prepend_rp'")
+
+    # verify x & y are dicts
+    _validate_mutable_mappings(x, y)
+
+    # to speed things up: if x is empty or equal to y, return y
+    # (this `if` can be remove without impact on the function
+    #  except performance)
+    if x == {} or x == y:
+        return y.copy()
+
+    # in the following we will copy elements from y to x, but
+    # we don't want to modify x, so we create a copy of it
+    x = x.copy()
+
+    # to speed things up: use dict.update if possible
+    # (this `if` can be remove without impact on the function
+    #  except performance)
+    if not recursive and list_merge == 'replace':
+        x.update(y)
+        return x
+
+    # insert each element of y in x, overriding the one in x
+    # (as y has higher priority)
+    # we copy elements from y to x instead of x to y because
+    # there is a high probability x will be the "default" dict the user
```

Comment:
Same here: I don't think the probability can be proven to be high, but that you're making an assumption here. So please state that you're making an assumption.

Reply:
done